src/Annotation.cpp

```diff
@@ -0,0 +1,500 @@
+
+    std::string fieldName =
+        dataflow::sliceString(targetChunks[1], targetChunks[1].find('(') + 1,
+                              targetChunks[1].find(')') - 1);
+    if (!dataflow::isValidCVariableName(fieldName)) {
+      logout("Annotation String Error 20: invalid annotation '"
+             << rawAnnotationString << "' FIELD argument '" << fieldName
+             << "' is not valid C variable name") return false;
+    }
+    // targetChunks[1] is "FIELD(str"
+
+    if (targetChunks[1].find(')') != targetChunks[1].size() - 1) {
+      logout("Annotation String Error 21: invalid annotation '"
+             << rawAnnotationString
+             << "' target is malformed; last character of '" << targetChunks[1]
+             << "' was not ')'") return false;
+    }
+    // targetChunks[1] is "FIELD(str)"
+  }
+  //* --- chunks[4] checker section close ---
+  if (chunks[5] != "methods") {
+    logout("Annotation String Error 22: invalid annotation '"
+           << rawAnnotationString
+           << "' missing keyword 'methods'/malformed methods "
+              "argument") return false;
+  }
+
+  if (chunks[6] != "=") {
+    logout("Annotation String Error 23: invalid annotation '"
+           << rawAnnotationString
+           << "' malformed methods argument") return false;
+  }
+
+  // chunk[7..end] should just be method args
+
+  std::string methodsString;
+  for (int i = 7; i < chunks.size(); i++) {
+
+    // removing potential "\0"'s in the string
+    chunks[i].erase(std::remove(chunks[i].begin(), chunks[i].end(), '\0'),
+                    chunks[i].end());
+
+    methodsString += chunks[i];
+
+    if (i != chunks.size() - 1) {
+      methodsString += " ";
+    }
+  }
+
+  if (!methodsArgumentIsCorrectlyFormatted(methodsString)) {
+    return false;
+  }
+
+  dataflow::removeWhitespace(methodsString);
+
+  std::vector<std::string> methodsVector =
+      dataflow::splitString(methodsString, ',');
+
+  for (std::string &method : methodsVector) {
+    if (!dataflow::isValidCVariableName(method)) {
+
+      logout("Annotation String Error 24: invalid annotation '"
+             << rawAnnotationString << "' method argument '" << method
+             << "' is not valid C string variable name") return false;
+    }
+  }
+
+  if (methodsVector.size() == 0) {
+    logout("Annotation String Error 26: invalid annotation '"
+           << rawAnnotationString << "' method argument empty") return false;
+  }
+
+  return true;
+}
+
+Annotation *generateAnnotation(const std::string &rawAnnotationString) {
+  if (!rawStringIsCorrectlyFormatted(rawAnnotationString)) {
+    return new ErrorAnnotation();
+  }
+
+  // TODO: issue an error if annotation is being re-defined
+
+  // we can assume string is correctly formatted
+
+  AnnotationType annoType;
+  std::vector<std::string> chunks =
+      dataflow::splitString(rawAnnotationString, ' ');
+
+  // raw annotation
+  // TOOL_CHECKER (Calls || MustCall) target = {FUNCTION ||
```

Comment:
This comment should be higher up in the file: above the validation code, definitely.